### PR TITLE
refactor: spec_finder.py

### PR DESCRIPTION
**Pull request by ghost**
## This PR

- applies ruff format
- adds missing docstrings for main
- uses %s formatting in logging functions
- adds encoding
- refactors extract_spec_map & parse_rust_files code from main
- refactors get_spec function
- adds logging config
- adds docstrings for _demarkdown & get_spec function
- improves message sentence

### Notes
Also, fixes parsing error for the specification.json

### CODE REVIEW

**Comment by gruebel** [comment]
hey @prnam thanks for the contribution, but there are too many unrelated changes in one PR, which makes it tedious to review. For example, don't apply a radical formatting, which changes mostly single quotes strings to double quotes. This is just a helper script, not really sure it needs proper logging or docstrings. Please create separate PRs for a subset of changes, then it can be reviewed better.

**Comment by beeme1mr** [comment]
@justinabrahms you may be interested in this one 😄

File: spec_finder.py [hunk 1]
```diff
@@ -88,18 +88,18 @@ def main(refresh_spec: bool = False, diff_output: bool = False, limit_numbers: O
...
                     logging.info('\n'.join([li for li in diff if not li.startswith(' ')]))
         else:
-            logging.info(f"{number} is defined in our tests, but couldn't find it in the spec")
+            logging.info("%s is defined in our tests, but couldn't find it in the spec", number)
 
     if missing:
         logging.info('In the spec, but not in our tests:')
         for m in sorted(missing):
-            logging.info(f"  {m}: {spec_map[m]}")
+            logging.info("%s: %s", m, spec_map[m])
```
Comment (justinabrahms) [inline spec_finder.py:121]: Just a note that this lost it's preceeding whitespace, which my fuzzy memory tells me wsa to help with visual parsing of the output.

File: spec_finder.py [hunk 2]
```diff
@@ -1,11 +1,15 @@
 #!/usr/bin/env python
+import logging
```
Comment (justinabrahms) [inline spec_finder.py]: I would personally not introduce the logging module. It's too heavy weight for CLI output, imo.

File: spec_finder.py [hunk 3]
```diff
@@ -34,74 +35,71 @@ def get_spec(force_refresh=False):
...
-                data = ''.join(f.readlines())
-
-            # if "#[spec" in data:
-            #     import pdb; pdb.set_trace()
-            for match in re.findall('#\[spec\((?P<innards>.*?)\)\]', data.replace('\n', ''), re.MULTILINE | re.DOTALL):
-                number = re.findall('number\s*=\s*"(.*?)"', match)[0]
-
-
-                if number in missing:
-                    missing.remove(number)
```
Comment (justinabrahms) [inline spec_finder.py:63]: I don't see this in the new code.

**Review by justinabrahms — COMMENTED** [review]
I like a lot of the changes in here. Unfortunately, I'm hoping that we can delete the spec_finder.py repo entirely in favor of a docker image that's used across languages. See #73. The new specfinder code is at https://github.com/open-feature/spec/tree/main/tools/specification_parser

**Pull request closed without merging**